### PR TITLE
[FLINK-14202][table][python] Optimize the execution plan for Python Calc when there is a condition

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request optimizes the Python UDF execution plan when there is a condition which contains Python UDFs by calculating the condition firstly and then check whether to call the other UDFs. This can eliminate the Python UDF execution for the input rows which could be filtered by the condition.*

## Brief change log

  - *Improves PythonScalarFunctionSplitRule to generate optimized plan.*
  - *Updates DataStreamPythonCalc to handle cases where condition is not null.*

## Verifying this change

This change is already covered by existing tests, such as PythonScalarFunctionSplitRuleTest, test_udf.py*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PythonCalcSplitRule.scala [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+    // fields of indexes greater than inputFieldsCount is the extracted functions and
+    // should be filtered as they are not from the original input
+    visitor.getFields.filter(_ < inputFieldsCount)
+  }
+
+  /**
+   * Whether to convert Python functions.
+   */
+  def convertPythonFunction(program: RexProgram): Boolean
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PythonCalcSplitRule.scala]: +1. renamed to isConvertPythonFunction

**Comment by dianfu** [comment]
@hequn8128 Thanks a lot for the review. Updated the PR.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PythonCalcSplitRule.scala [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalCalc
+import org.apache.flink.table.planner.plan.utils.PythonUtil.containsFunctionOf
+import org.apache.flink.table.planner.plan.utils.{InputRefVisitor, RexDefaultVisitor}
+
+import scala.collection.JavaConverters._
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+/**
+  * Rule that splits [[FlinkLogicalCalc]] into multiple [[FlinkLogicalCalc]]s.
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/PythonCalcSplitRule.scala]: Base Rule that splits [[FlinkLogicalCalc]] into multiple [[FlinkLogicalCalc]]s. It is mainly to ensure that each [[FlinkLogicalCalc]] only contains java/scala or python scalar functions?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonCalc.scala [hunk 3]
```diff
@@ -55,18 +53,15 @@ class StreamExecPythonCalc(
...
+    val ret = createPythonOneInputTransformation(
       inputTransform,
-      inputsContainSingleton(),
       calcProgram,
-      getRelDetailedDescription,
-      config,
-      ctx,
-      cluster,
-      getRowType,
       "StreamExecCalc")
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonCalc.scala]: "StreamExecCalc" => "StreamExecPythonCalc"

**Comment by dianfu** [comment]
@hequn8128 Thanks a lot for the review. Updated.

**Comment by hequn8128** [comment]
@dianfu LGTM. Will merge once Travis passed.

**Pull request closed without merging**